convert DxfPoint and DxfVector to structs

brettfo opened the pull request:
There are multiple issues with having `DxfPoint` and `DxfVector` be `class`es instead of `struct`s and this PR does the appropriate conversion by copying the implementation from IxMilia/BCad ([here](https://github.com/IxMilia/BCad/blob/1a9aef54176fd965eeb8262d9004f6649e5eb60f/src/IxMilia.BCad.Core/Point.cs) and [here](https://github.com/IxMilia/BCad/blob/1a9aef54176fd965eeb8262d9004f6649e5eb60f/src/IxMilia.BCad.Core/Vector.cs)).

Since this is a breaking change I'm interested in feedback, particularly from @Idaps and @nzain because taking this change would make it much easier to add bounding box calculations to entities.

Since `struct`s are immutable it gets a little ugly internally when only a single component of the point/vector needs to be updated, but all of that ugliness (the `.UpdateX()`, etc. methods) is hidden from the public API.  In short, this `line.P1.X = newValue` is changed to this `line.P1 = line.P1.UpdateX(newValue)`.

nzain commented:
will look into it asap - but blocked right now by other issues.

nzain commented:
Without looking at the PR: in https://github.com/IxMilia/Dxf/issues/75#issuecomment-357382131 you said

> In short, since DxfPoint is a class it can be null which makes semi-auto-generating GetBoundingBox() difficult without requiring jumping through a lot of hoops.

Using a struct will make this even more complicated, because you cannot check for `null`. Instead, a struct will always be filled with zeros, e.g. a point at (0,0) which is even more tricky to distinguish from the real point at (0,0). Avoid `NaN` if you can, makes life really complicated.

You want to use structs for small objects to avoid GC pressure, when millions of them are created and dismissed quickly. Furthermore, you have to avoid deep call stacks, because you struct usually gets copied for every level. I don't think this project needs better performance. Personally, I use structs only when streaming gigabytes of 3D laser scan points through some of my processing pipelines.

Regarding 'immutable' objects: this is always beneficial, for both struct/class. Avoids a bunch of problems but doesn't work well with "change one property of an object". Right now, I don't care because I don't modify points or vectors (iirc).

66 changed files... will briefly look at the PR now.

nzain commented on src/IxMilia.Dxf/DxfPoint.cs:
```diff
@@ -1,94 +1,101 @@
+                return true;
+            if (((object)p1 == null) || ((object)p2 == null))
+                return false;
+            return p1.X == p2.X && p1.Y == p2.Y && p1.Z == p2.Z;
```
This looks like class equality, should be updated for structs.

nzain commented on src/IxMilia.Dxf/DxfPoint.cs:
```diff
@@ -1,94 +1,101 @@
-        public static DxfVector XAxis
+        // the following methods are only used to allow setting individual x/y/z values in the auto-generated readers
+
+        internal DxfPoint UpdateX(double x)
```
The name `UpdateX` implies that the "object" is updated, which is not the case. You might ignore the return value and call the method - bug. Some people would call this method `UpdatedX` or `GetUpdatedX` to highlight that the method does not alter the object. Confusing in any case, I prefer constructors only and avoid update methods on immutable objects. However, this is personal oppinion :-)

nzain commented:
Ok, as far as I could see you default to zero anyways, so a struct would simplify this. Just be really careful with `null` checks, they might make sense with a class. I took a closer look at `DxfPoint` and added some comments there.

brettfo commented:
Thanks for the quick feedback.  I've fixed the equality checks to be more struct-like (I think that `ReferenceEquals()` stuff was from way back when that other Point was a class instead of a struct).  I've also renamed `UpdateX()` to `WithUpdatedX()` which isn't great, but I think conveys the semantics a little better, even if it is more verbose.
